syntax = "proto3";

package CoffeeMachine;

message TemperatureControl {
    float setpoint = 1;  // Desired boiler temperature in Celsius.
}

message PinControl {
    enum Pin {
        PUMP = 0;
        BOILER = 1;
    }
    Pin pin = 1;         // Pin to control.
    bool state = 2;      // True = ON, False = OFF.
}

message TemperatureRead {
    float temperature = 1;  // Current boiler temperature in Celsius.
}

message PinRead {
    enum Pin {
        PUMP = 0;
        BOILER = 1;
    }
    Pin pin = 1;            // Pin to read.
    bool state = 2;         // True = ON, False = OFF.
}

message PIDAutotune {
}

message Ping {
    bool is_alive = 1;   // True if the UI is responsive.
}

message Error {
    int32 error_code = 1; // Numeric error code
    string error_message = 2; // Human-readable error message
}

message Command {
    oneof cmd {
        TemperatureControl temp_control = 1;
        PinControl pin_control = 2;
        TemperatureRead temp_read = 3;
        PinRead pin_read = 4;
        PIDAutotune pid_autotune = 5;
        Ping ping = 6;
        Error error = 7;
    }
}